File: lib/src/main/java/net/imoya/android/preference/controller/StringPreferenceEditor.java
package net.imoya.android.preference.controller;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Parcel;
import android.os.Parcelable;
import android.text.InputType;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import net.imoya.android.dialog.DialogBase;
import net.imoya.android.dialog.TextInputDialog;
import net.imoya.android.preference.view.SingleValuePreferenceView;
import net.imoya.android.preference.view.StringPreferenceView;

/**
 * 文字列設定値編集コントローラ
 * <p/>
 * {@link StringPreferenceView} と組み合わせて使用することを想定しています。
 */
public class StringPreferenceEditor extends DialogPreferenceEditor {
    /**
     * 状態オブジェクト
     */
    private static class State extends DialogPreferenceEditor.State {
        /**
         * 入力欄のヒント文字列
         */
        private String hint;
        /**
         * 入力文字タイプ
         */
        private int inputType = InputType.TYPE_CLASS_TEXT;
        /**
         * 最大入力可能文字数
         */
        private int maxLength = Integer.MAX_VALUE;

        private State() {
            super();
        }

        private State(Parcel parcel) {
            super(parcel);
            this.hint = parcel.readString();
            this.inputType = parcel.readInt();
            this.maxLength = parcel.readInt();
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            super.writeToParcel(dest, flags);
            dest.writeString(this.hint);
            dest.writeInt(this.inputType);
            dest.writeInt(this.maxLength);
        }

        /**
         * {@link Parcelable} 対応用 {@link Creator}
         */
        public static final Creator<State> CREATOR = new Creator<State>() {
            /**
             * {@link Parcel} の内容を保持するオブジェクトを生成して返します。
             *
             * @param parcel {@link Parcel}
             * @return {@link Parcel} の内容を保持するオブジェクト
             */
            @Override
            public State createFromParcel(Parcel parcel) {
                return new State(parcel);
            }

            /**
             * オブジェクトの配列を生成して返します。
             *
             * @param size 配列のサイズ
             * @return 配列
             */
            @Override
            public State[] newArray(int size) {
                return new State[size];
            }
        };
    }

    // private static final String TAG = "StringPreferenceEditor";

    /**
     * コンストラクタ
     *
     * @param fragment    設定項目ビューを表示する {@link Fragment}
     * @param preferences 設定値が保存される {@link SharedPreferences}
     * @param requestCode 設定ダイアログの識別に使用するリクエストコード
     * @param <T>         fragment は {@link DialogBase.Listener} を実装した {@link Fragment} であること
     */
    public <T extends Fragment & DialogBase.Listener> StringPreferenceEditor(
            T fragment, SharedPreferences preferences, int requestCode) {
        super(fragment, preferences, requestCode);
    }

    public void setInputType(int inputType) {
        ((State) this.state).inputType = inputType;
    }

    public void setHint(String hint) {
        ((State) this.state).hint = hint;
    }

    @Override
    protected DialogPreferenceEditor.State createState() {
        return new State();
    }

    @Override
    protected void setupState(@NonNull SingleValuePreferenceView view) {
        super.setupState(view);

        final StringPreferenceView prefView = (StringPreferenceView) view;
        ((State) this.state).maxLength = prefView.getMaxLength();
    }

    @Override
    protected void showDialog(SingleValuePreferenceView view) {
        new TextInputDialog.Builder(
                new DialogBase.BuilderParentFragment<>(
                        (Fragment & DialogBase.Listener) this.fragment
                ),
                this.getRequestCode())
                .setTitle(view.getTitle())
                .setInputType(((State) this.state).inputType)
                .setMaxLength(((State) this.state).maxLength)
                .setHint(((State) this.state).hint)
                .setText(this.preferences.getString(this.state.key, ""))
                .setTag(view.getPreferenceKey())
                .show();
    }

    @Override
    protected void saveInput(int resultCode, @NonNull Intent data) {
        final String text = data.getStringExtra(TextInputDialog.EXTRA_KEY_INPUT_VALUE);
        this.preferences.edit().putString(this.state.key, text).apply();
    }
}

File: lib/src/main/java/net/imoya/android/preference/controller/StringPreferenceEditor.kt
/*
 * Copyright (C) 2022 IceImo-P
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.imoya.android.preference.controller

import android.content.Intent
import android.content.SharedPreferences
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import android.text.InputType
import net.imoya.android.dialog.DialogParent
import net.imoya.android.dialog.InputDialog
import net.imoya.android.dialog.TextInputDialog
import net.imoya.android.preference.Constants
import net.imoya.android.preference.view.PreferenceView
import net.imoya.android.preference.view.StringPreferenceView

/**
 * 文字列設定値編集コントローラ
 *
 * [StringPreferenceView] と組み合わせて使用することを想定しています。
 */
@Suppress("unused")
open class StringPreferenceEditor(
    /**
     * 設定ダイアログの親画面
     */
    parent: DialogParent? = null,
    /**
     * 設定値が保存される [SharedPreferences]
     */
    preferences: SharedPreferences? = null,
    /**
     * 設定ダイアログの識別に使用するリクエストコード
     */
    requestCode: Int = Constants.DEFAULT_REQUEST_CODE
) : DialogPreferenceEditor(parent, preferences, requestCode) {
    /**
     * 状態オブジェクト
     */
    protected open class State : PreferenceEditor.State {
        /**
         * 入力欄のヒント文字列
         */
        var hint: String? = null

        /**
         * 入力文字タイプ
         */
        var inputType = InputType.TYPE_CLASS_TEXT

        /**
         * 最大入力可能文字数
         */
        var maxLength = Int.MAX_VALUE

        constructor() : super()

        /**
         * [Parcel] の内容で初期化するコンストラクタ
         *
         * @param parcel [Parcel]
         */
        protected constructor(parcel: Parcel) : super(parcel) {
            hint = parcel.readString()
            inputType = parcel.readInt()
            maxLength = parcel.readInt()
        }

        override fun writeToParcel(dest: Parcel, flags: Int) {
            super.writeToParcel(dest, flags)
            dest.writeString(hint)
            dest.writeInt(inputType)
            dest.writeInt(maxLength)
        }

        companion object {
            /**
             * [Parcelable] 対応用 [Creator]
             */
            @JvmField
            val CREATOR: Creator<State> = object : Creator<State> {
                /**
                 * [Parcel] の内容を保持するオブジェクトを生成して返します。
                 *
                 * @param parcel [Parcel]
                 * @return [Parcel] の内容を保持するオブジェクト
                 */
                override fun createFromParcel(parcel: Parcel): State {
                    return State(parcel)
                }

                /**
                 * オブジェクトの配列を生成して返します。
                 *
                 * @param size 配列のサイズ
                 * @return 配列
                 */
                override fun newArray(size: Int): Array<State?> {
                    return arrayOfNulls(size)
                }
            }
        }
    }

    fun setInputType(inputType: Int) {
        (state as State).inputType = inputType
    }

    fun setHint(hint: String?) {
        (state as State).hint = hint
    }

    override fun isCompatibleView(view: PreferenceView): Boolean {
        return view is StringPreferenceView
    }

    override fun createState(): PreferenceEditor.State {
        return State()
    }

    override fun setupState(view: PreferenceView) {
        super.setupState(view)
        if (view !is StringPreferenceView) {
            throw IllegalArgumentException("View must be StringPreferenceView")
        }
        (state as State).maxLength = view.maxLength
    }

    override fun showDialog(view: PreferenceView) {
        val currentPreferences = checkPreferences()
        val currentParent = checkParent()
        val key = checkKey()
        if (view !is StringPreferenceView) {
            throw IllegalArgumentException("View must be StringPreferenceView")
        }
        checkAndWarnRequestCode()
        val currentState = state as State

        TextInputDialog.Builder(currentParent, requestCode)
            .setTitle(view.title ?: "")
            .setInputType(currentState.inputType)
            .setMaxLength(currentState.maxLength)
            .setHint(currentState.hint)
            .setText(currentPreferences.getString(key, "") ?: "")
            .setTag(key)
            .show()
    }

    override fun saveInput(resultCode: Int, data: Intent) {
        val currentPreferences = checkPreferences()
        val key = checkKey()
        val value = data.getStringExtra(InputDialog.EXTRA_KEY_INPUT_VALUE)
        currentPreferences.edit().putString(key, value).apply()
    }

    companion object {
        /**
         * Tag for log
         */
        private const val TAG = "StringPreferenceEditor"
    }
}